apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/AckError.cs

```diff
@@ -0,0 +1,152 @@
+    /// </summary>
+    /// <param name="rpcException">The underlying <see cref="RpcException"/>.</param>
+    /// <param name="ids">The ids for which the Acknowledge/ModifyAckDeadline RPC was called.</param>
+    /// <returns>The populated instance of <see cref="AckError"/> containing temporary and permanent failures.</returns>
+    internal static AckError ForRpcException(RpcException rpcException, IEnumerable<string> ids)
+    {
+        GaxPreconditions.CheckNotNull(rpcException, nameof(rpcException));
+        GaxPreconditions.CheckNotNull(ids, nameof(ids));
+
+        // Check the gRPC StatusCode. For certain StatusCodes, we need to retry all ids.            
+        bool retryAll = ShouldRetryAll(rpcException);
+        if (retryAll)
+        {
+            // Return all ids in temporary failures and empty permanent failures, so that the caller can retry the whole temporary batch.
+            return new AckError(ids.ToDictionary(key => key, value => rpcException.Status.StatusCode.ToString()), Enumerable.Empty<KeyValuePair<string, string>>());
+        }
+
+        // We can have complete or partial failure. We classify the error as temporary or permanent
+        // based on the error message starting from TRANSIENT_FAILURE_ or PERMANENT_FAILURE_ (which is ugly but that's the way it is)
+        // and return:
+        // the temporary errors in temporaryErrors collection and
+        // the permanent errors in permanentErrors collection.
+        // For exactly once, the metadata of the rpcException contains the error details. A sample metadata is shown below:
+        // {
+        //    "ack_id_1": "PERMANENT_FAILURE_INVALID_ACK_ID",
+        //    "ack_id_2": "PERMANENT_FAILURE_INVALID_ACK_ID",
+        //    "ack_id_3": "TRANSIENT_FAILURE_UNORDERED_ACK_ID"
+        // }
+        var metadata = rpcException.GetErrorInfo()?.Metadata ?? Enumerable.Empty<KeyValuePair<string, string>>();
+        var temporaryErrors = metadata.Where(j => j.Value.StartsWith("TRANSIENT_FAILURE_", StringComparison.Ordinal));
+        var permanentErrors = metadata.Where(j => j.Value.StartsWith("PERMANENT_FAILURE_", StringComparison.Ordinal));
+
+        if (temporaryErrors.Any() || permanentErrors.Any())
+        {
+            // If temporary or permanent error messages are present, it means the entire request hasn't failed with a permanent error.
+            return new AckError(temporaryErrors, permanentErrors);
+        }
+        else
+        {
+            // Now, if temporary or permanent error messages are *not* present, it means the entire request has failed with a permanent error.
+            // It may be due to a gRPC error like a failed precondition or a permission denied error or other permanent error.
+            return new AckError(temporaryErrors, ids.ToDictionary(key => key, value => rpcException.Status.StatusCode.ToString()));
```

Comment:
I know `temporaryErrors` will be empty at this point. But passing in a variable means we can to trace the logic of the variable and deduce this fact. Perhaps just pass in an empty container instead, just to make it easier to follow?